agent:overlay: Add underlayProtocol 'Auto' for automatic underlay selection

smagnani96 opened the pull request:
Superseeds https://github.com/cilium/cilium/pull/42861.

Please refer to commit messages:

1. Add new default `underlayProtocol: "auto"`. This will allow Cilium to automatically pick an enabled underlay as protocol (IPv4 first, otherwise IPv6). The change is transparent to all users who do not explicitly define `underlayProtocol`.
2. Prohibit setting a disabled `underlayProtocol`. Prior to this, Cilium did not throw any error, but it was using the wrong TunnelOverhead while computing routeMTU for pods in pkg/mtu.

Fixes: #40877.

pchaigno commented on pkg/datapath/tunnel/tunnel.go:
```diff
@@ -88,7 +93,16 @@ func newConfig(in newConfigIn) (Config, error) {
+		}
+	case IPv6:
+		if !in.DaemonCfg.EnableIPv6 {
+			in.Logger.Info("Selected underlay IPv6 is disabled, switching to IPv4.")
```
Should this be a warning so people can update their configmap accordingly and we don't switch on every restart?

smagnani96 replied:
Are you suggesting to KEEP the switching logic but log this as Warning? If so, I agree. Otherwise I don't, I'd still want something to happen to avoid wrong MTU computation. Agent failure would've been my 1st option, but that would cause problems to users with a valid IPV6-only+NativeRouting config (see discussion https://github.com/cilium/cilium/pull/42861#issuecomment-3551959763).

pchaigno replied:
> Are you suggesting to KEEP the switching logic but log this as Warning?

Yes.

smagnani96 replied:
Done. Thanks for clarifying!

smagnani96 replied:
Changing log to Warn makes `ci-clustermesh` [fail](https://github.com/cilium/cilium/actions/runs/20858980298/job/59933379371).
That of course happened in [NativeRoutingIPv6](https://github.com/cilium/cilium/blob/pr/smagnani96/overlay-switch-underlay/.github/workflows/conformance-clustermesh.yaml#L164-L176) where tunneling is disabled. We should still specify the correct underlay to not receive the warning.
That warning makes our CI to fail.
Added a new commit to this PR.

joestringer commented on pkg/datapath/tunnel/tunnel.go:
```diff
@@ -88,7 +93,16 @@ func newConfig(in newConfigIn) (Config, error) {
-	case IPv4, IPv6:
+	case IPv4:
+		if !in.DaemonCfg.EnableIPv4 {
+			in.Logger.Warn("Selected underlay IPv4 is disabled, switching to IPv6.")
+			in.Cfg.UnderlayProtocol = string(IPv6)
+		}
```
Did you consider just returning an error here?

As a general configuration problem, when the user explicitly asks for option A && option B, we should validate the combination and either accept or reject it. However if the user doesn't explicitly ask for this combination, it is reasonable to autodetect an option.

The concern I would have with this pattern specifically is if the user explicitly selected `--helm-set=underlayProtocol=ipv6` and did not enable IPv6, they might continue to operate Cilium assuming that the underlay uses IPv6 (as they have configured) and only later realize that Cilium overrode their configuration to work differently. That is surprising behavior, and generally speaking users don't love to be surprised.

smagnani96 replied:
My favorite option would be to return error, so that users are well aware of what needs to be set/changed.
However, in https://github.com/cilium/cilium/pull/42861#issuecomment-3551959763 Julian had the following concern:

> Just saying that IPv6-only with native-routing has been a valid config for a long time, and all those folks would need to fix-up their config now (the underlay-protocol config option only exists since v1.18)

And that's because the default value would be `underlayProtocol: ipv4`, even though these users are in IPv6-only clusters. It's also a bit controversial to me that users need to specify the correct underlay even though they're in NativeRouting, but apparently we still do use it in `pkg/mtu` [here](https://github.com/cilium/cilium/blob/main/pkg/mtu/cell.go#L88-L96) and [here](https://github.com/cilium/cilium/blob/main/pkg/mtu/mtu.go#L91-L95).

I'm super open for any kind of suggestions. Don't know what's the preferred way to go in such cases 😃

joestringer replied:
Could we have an `underlayProtocol: auto` that picks ipv4 or ipv6 (in that order), based on which protocols are enabled in the config? If we then set that as the default, it should generally have the right behavior, but then for dual-stack environments the user could choose to opt in to a specific underlay. Then we could still error out if explicitly configured the wrong way.

smagnani96 replied:
The `auto` solution sounds good, don't see any issue ATM. Tomorrow I'll give it a try, thanks Joe 💯

smagnani96 replied:
My "Tomorrow" has a two-weeks lag 😁 
Just updated the PR with the `auto` solution.

smagnani96 committed:
tunnel: Error when providing a disabled underlayProtocol

Prior to this commit, it was still possible to specify a disabled
underlayProtocol in the Cilium config. With this, we explicitly error out
and inform the user that they cannot provide a disabled underlayProtocol.
This makes the configuration more explicit and avoids confusion.

An example is in IPv6-only clusters with IPSec enabled and tunnel underlay
left to IPv4 (disabled by agent config). In this case, we would compute
pods routeMTU accounting for IPv4 tunnel overhead (50B) rather than IPv6
(70B), which is incorrect. That led to connectivity issues and fragmentation.

See https://github.com/cilium/cilium/issues/23917.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

pchaigno reviewed (approved):
In general, I'm a bit wary of `auto` flags as it tends to create an unstable environment for users, but in this case, it's only dependent on other config flags, so I think we should be fine.

qmonnet reviewed (approved):
Ack for doc change, thanks!

smagnani96 committed:
tunnel: Add support for automatic underlay selection

This commit introduces a new `underlayProtocol: "auto"` option in the
tunnel configuration, and this will be from this commit onwards the
default option unless specified otherwise.
With this option, the system will automatically detect and select the
appropriate underlay protocol (IPv4 or IPv6) based on the daemon config,
specifically the EnableIPv4 and EnableIPv6 settings. If both IPv4 and IPv6
are enabled, IPv4 will be preferred as we were already doing.

Prior to this, we were assuming IPv4 underlay unless explicitly specified otherwise,
also in case of tunneling disabled. However, that leads to MTU miscalculations
in pkg/mtu in case of IPv6-only clusters with tunneling disabled. The resulting
MTU computed would account for an IPv4 underlay overhead, while it should've
considered the IPv6 overhead. Simply prohibiting the user from specifying
a disabled underlay protocol (e.g., underlayProtocol: "ipv4" in an IPv6-only cluster)
was not the best option, given it would require all users to update their
Cilium config, including those running legit IPv6-only native clusters.

This commit fixes this behavior transparently to all users that were not
manually/explicitly setting the underlayProtocol option during installation.

This commit does not fix the case in which a disabled underlayProtocol is
specified (e.g., underlayProtocol: "ipv4" in an IPv6-only cluster). That will
be addressed in a follow-up commit.

This commit updates helm chart, agent default underlayProtocol, and
documentation accordingly.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>
